cassandra/src/main/java/org/hibernate/ogm/datastore/cassandra/impl/CassandraDatastoreProvider.java

```diff
@@ -0,0 +1,266 @@
+ *
+ * @author Jonathan Halliday
+ */
+public class CassandraDatastoreProvider extends BaseDatastoreProvider
+		implements Startable, Stoppable, ServiceRegistryAwareService, Configurable {
+
+	private static final Log log = LoggerFactory.getLogger();
+
+	private ServiceRegistryImplementor serviceRegistry;
+
+	private CassandraConfiguration config;
+
+	private Cluster cluster;
+	private Session session;
+	private CassandraSequenceHandler sequenceHandler;
+
+	private final Set<String> entityTableNames = new HashSet<String>();
+
+	private final Map<String, Table> metaDataCache = new HashMap<String, Table>();
+
+	public void setTableMetadata(String name, Table table) {
+		metaDataCache.put( name, table );
+	}
+
+
+	@Override
+	public Class<? extends SchemaDefiner> getSchemaDefinerType() {
+		return CassandraSchemaDefiner.class;
+	}
+
+	public CassandraSequenceHandler getSequenceHandler() {
+		return sequenceHandler;
+	}
+
+	@Override
+	public void configure(Map configurationValues) {
+		OptionsService optionsService = serviceRegistry.getService( OptionsService.class );
+		ClassLoaderService classLoaderService = serviceRegistry.getService( ClassLoaderService.class );
+		ConfigurationPropertyReader propertyReader = new ConfigurationPropertyReader(
+				configurationValues,
+				classLoaderService
+		);
+		config = new CassandraConfiguration( propertyReader, optionsService.context().getGlobalOptions() );
+	}
+
+	public Session getSession() {
+		return session;
+	}
+
+	public Map<String, Table> getMetaDataCache() {
+		return metaDataCache;
+	}
+
+	@Override
+	public Class<? extends GridDialect> getDefaultDialect() {
+		return CassandraDialect.class;
+	}
+
+	@Override
+	public void injectServices(ServiceRegistryImplementor serviceRegistry) {
+		this.serviceRegistry = serviceRegistry;
+	}
+
+	@Override
+	public void start() {
+		if ( cluster == null ) {
+			try {
+				log.connectingToCassandra( config.getHost(), config.getPort() );
+
+				cluster = new Cluster.Builder()
+						.addContactPoint( config.getHost() )
+						.withPort( config.getPort() )
+						.build();
+
+				Session bootstrapSession = cluster.connect();
+				bootstrapSession.execute(
+						"CREATE KEYSPACE IF NOT EXISTS " + config.getDatabaseName() +
+								" WITH replication = {'class': 'SimpleStrategy', 'replication_factor': 1 }"
+				);
+				bootstrapSession.close();
+
+				session = cluster.connect( config.getDatabaseName() );
+
+				sequenceHandler = new CassandraSequenceHandler(this);
+			}
+			catch (RuntimeException e) {
+				throw log.unableToInitializeCassandra( e );
+			}
+		}
+	}
+
+	@Override
+	public void stop() {
+		log.disconnectingFromCassandra();
+		session.close();
+		session = null;
+		cluster.close();
+		cluster = null;
+		sequenceHandler = null;
+	}
+
+	public void removeKeyspace() {
+		Session bootstrapSession = cluster.connect();
+		bootstrapSession.execute( "DROP KEYSPACE " + config.getDatabaseName() );
+		bootstrapSession.close();
+	}
+
+	// test harness
```

Comment:
methods moved. metaDataCache has to be somewhere - I left it on the datastore because the alternative means linking production code to the test helper i.e. the table creation code would have to populate the cache in the test helper.